rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/monitoring/views/CompactFlowPanel.java

```diff
@@ -0,0 +1,430 @@
+                        if (hierarchyHost.getHost().getHostName().equals(node.getHostName())) {
+                            ++index;
+
+                            final Iterator<NodeSource.Host.Node> nodeIterator = hierarchyHost.getNodes().iterator();
+                            while (nodeIterator.hasNext()) {
+                                final NodeSource.Host.Node existing = nodeIterator.next();
+                                if (existing.equals(node)) {
+                                    nodeIterator.remove();
+                                    hierarchyNodeSource.decrementTiles();
+                                    hierarchyHost.decrementTiles();
+                                    this.remove(index);
+
+                                    // remove dangling host
+                                    if (hierarchyHost.getNodes().isEmpty()) {
+                                        hostIterator.remove();
+                                        hierarchyNodeSource.decrementTiles();
+                                        this.remove(index - 1);
+                                    }
+                                    return;
+                                } else {
+                                    ++index;
+                                }
+                            }
+
+                        } else {
+                            index += hierarchyHost.getTiles();
+                        }
+                    }
+
+                }
+
+            } else {
+                index += hierarchyNodeSource.getTiles();
+            }
+        }
+
+    }
+
+    public Optional<Integer> indexOf(NodeSource ns) {
+        int index = 0;
+        for (HierarchyNodeSource hierarchyNodeSource : model) {
+            if (hierarchyNodeSource.getNodeSource().getSourceName().equals(ns.getSourceName())) {
+                return Optional.of(index);
+            } else {
+                index += hierarchyNodeSource.getTiles();
+            }
+        }
+        return Optional.empty();
+    }
+
+    public Optional<Integer> indexOf(NodeSource.Host host) {
+        int index = 0;
+        for (HierarchyNodeSource hierarchyNodeSource : model) {
+            if (hierarchyNodeSource.getNodeSource().getSourceName().equals(host.getSourceName())) {
+                ++index;
+                index += hierarchyNodeSource.getDeploying().size();
+                for (HierarchyHost hierarchyHost : hierarchyNodeSource.getHosts()) {
+                    if (hierarchyHost.getHost().getHostName().equals(host.getHostName())) {
+                        return Optional.of(index);
+                    } else {
+                        index += hierarchyHost.getTiles();
+                    }
+                }
+
+            } else {
+                index += hierarchyNodeSource.getTiles();
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    public Optional<Integer> indexOf(NodeSource.Host.Node node) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 42 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)